include/broker/zeek.hh

```diff
@@ -105,6 +130,104 @@ public:
     return get<std::string>(get<vector>(as_vector()[2])[0]);
   }
 
+  /// Support iteration with structured binding.
+  class metadata_iterator {
+    using iterator_category = std::forward_iterator_tag;
+    using difference_type = std::ptrdiff_t;
+    using value_type = std::pair<count, const data&>;
+    using pointer = value_type*;
+    using reference = value_type&;
+
+  public:
+    metadata_iterator(const broker::data* ptr) : ptr_(ptr) {}
+
+    const std::pair<count, const data&> operator*() const {
+      auto entry_ptr = get_if<vector>(*ptr_);
+      BROKER_ASSERT(entry_ptr && entry_ptr->size() == 2);
+      const auto& entry = *entry_ptr;
+      return {get<count>(entry[0]), entry[1]};
+    }
+
+    metadata_iterator operator++() {
+      ++ptr_;
+      return *this;
+    }
+
+    metadata_iterator operator++(int) {
+      auto tmp = *this;
+      ++(*this);
+      return tmp;
+    }
+
+    bool operator!=(metadata_iterator& other) {
+      return ptr_ != other.ptr_;
+    }
+
+  private:
+    const broker::data* ptr_;
+  };
+
+  /// Supports iteration over metadata
+  class metadata_wrapper {
+  public:
+    metadata_wrapper(const vector* v) : v_(v) {}
+
+    metadata_iterator begin() const {
+      return metadata_iterator{v_->data()};
+    }
+
+    metadata_iterator end() const {
+      return metadata_iterator{v_->data() + v_->size()};
+    }
+
+    const data* value(Metadata key) const {
+      return value(static_cast<count>(key));
+    }
+
+    const data* value(count key) const {
+      for (const auto& [k, v] : *this) {
+        if (k == key)
+          return &v;
+      }
+      return nullptr;
+    }
+
+    const auto& as_vector() const {
+      return *v_;
+    }
+
+  private:
+    const vector* v_;
+  };
+
+  const std::optional<metadata_wrapper> metadata() const {
```

Comment:
You make a point of returning `optional<metadata_wrapper>`, indicating the wrapper may not be null/empty. I can kinda see your reasoning for this, but from a user perspective, this wrapper just represents a sequence of elements. I don't think we'd ever have an actual need for distinguishing between `nullopt` and `empty` here (plus, there's also that weird `const` on a value return type again).

So, what I'd do instead is to allow constructing the wrapper from `nullptr`. The only member that has to go is `as_vector()`, which just looks like an unnecessary convenience to me (but it does lock the class into asserting a non-null `v_` member.

```suggestion
  metadata_wrapper metadata() const {
```

Reply:
Yep, I'll remove it. The Python bindings were the only user, but I'll work-around it then.

>  (plus, there's also that weird const on a value return type again).

I understand what's wrong there now :-/

Reply:
> Yep, I'll remove it. The Python bindings were the only user, but I'll work-around it then.

I don't think you'll like the work-around - accessor to the raw vector - but the optional is gone now.

I've also removed the `metadata()` overloads from MetadataWrapper because now once can just call `ev.metadata().value(1234)` to do the same.